### ISSUE TITLE
nn.Helper layer

### ISSUE THREAD

**Issue by blackyang** [torch/nn#1155]
Hi, is there any layers that can help debugging? Inspired by dpnn's `nn.PrintSize()` layer, I have a `nn.Helper()` layer

```
local Helper, parent = torch.class('nn.Helper', 'nn.Module')

function Helper:__init(func)
   parent.__init(self)
   self.func = func
end

function Helper:updateOutput(input)
   self.output = input
   func(input, self.output, gradInput, self.gradOutput)
   return self.output
end


function Helper:updateGradInput(input, gradOutput)
   self.gradInput = gradOutput
   func(input, self.output, gradInput, self.gradOutput)
   return self.gradInput
end
```

I can create a PR if this `nn.Helper` layer is helpful

**Comment by richardassar** [comment]
I've modified this to fix some errors and included a boolean param to signal whether we're in `updateOutput` or `updateGradInput`. 

There's no sense in passing `self.gradInput` during `updateOutput` as the values will be stale. A neater alternative might be to accept two funcs in the constructor, one for each phase.

`local Helper, parent = torch.class('nn.Helper', 'nn.Module')

function Helper:__init(func)
   parent.__init(self)
   self.func = func
end

function Helper:updateOutput(input)
   self.output = input
   self.func(input, self.gradInput, false)
   return self.output
end

function Helper:updateGradInput(input, gradOutput)
   self.gradInput = gradOutput
   self.func(input, self.gradInput, true)
   return self.gradInput
end`

**Comment by blackyang** [comment]
Hi @richardassar , thanks for following up! I just quickly wrote the implementation of `nn.Helper` using github's input text block, so there could be errors haha

My original thought is that we give `func` full access to all interval variables, and let programmers to decide what to do within `func`. So it could be something like:
```
self.func(self)
```

Using two funcs in the constructor is also a good idea

**Comment by richardassar** [comment]
@blackyang np, I had the need to inspect some intermediate values so it was a useful starting point. Thanks.

Passing `self` to the lambdas makes this neater.

For multiple instantiation, avoiding repetition you could return a constructor instead e.g.

```
Helper = nn.Helper(function(input) 
    --[[ debug some stuff ]]--
end, function(gradInput) 
    --[[ debug some stuff ]]--
end)
```

then have...

```
nn.Sequential()
    :add(Helper("Conv layer 1"))
    :add(...)
    ...
    :add(Helper("Before residual block"))
```

which would pass the label into the lambda.

Without wanting to over-complicate matters it might be nice to have a helper function that takes in a table of functions that fire before and/or after certain methods via swizzling e.g.

`nn.Helper({updateOutput = {function before(input) ... end, function after(input) ... end}, updateGradInput = {function before(input, gradOutput) ... end, function after(input, gradOutput) ... end}}, module)`

This could make use of the pattern above also.
